rfcs/2021-11-03-9862-ingest-apm-stats-along-traces-in-dd-agent-source.md

```diff
@@ -0,0 +1,292 @@
+each instrumented resource (i.e. a given piece of code) that are aggregated over time (based on 100% of the traces
+received). Those APM stats are very important as they can highlight code hot spots and ease aggregation. Also as an
+integral part of the Datadog APM product, APM stats should be handled by Vector.
+
+[trace-support-pr]: https://github.com/vectordotdev/vector/pull/9634
+
+## Context
+
+Traces are new in Vector, initial support is focusing on traces coming from the Datadog Agent, the
+[RFC][trace-support-pr] discussing about traces describes how those traces will be handled in Vector and highlighted the
+need to support APM stats.
+
+APM stats are encoded using protobuf, the current schema is available in the [datadog-agent][apm-stats-proto]
+repository. APM stats may be computed by the tracing lib in some cases or most of the time by the trace agent, but once
+they are emitted by the `trace-agent` there is no difference except a [boolean value][client-side-marker] that indicates
+where it was computed. They are sent by the `trace-agent` to the same endpoint as trace payloads, only the path differ,
+allowing easy discrimination between trace & APM stats payloads.
+
+Those stats are computed by a component named [concentrator] in the trace-agent. There is a [dedicated
+path][client-stats-aggregator] for APM stats that comes directly from tracing libraries. But ultimately they flow
+through datadog with the same [sending code][stats-writer].
+
+If we details a bit the actual content of a stats payload, it is an aggregate of
+[`ClientGroupedStats`][client-grouped-stats-proto]:
+
+```protobuf
+string service = 1;
+string name = 2;
+string resource = 3;
+uint32 HTTP_status_code = 4;
+string type = 5;
+string DB_type = 6; // db_type might be used in the future to help in the obfuscation step
+uint64 hits = 7; // count of all spans aggregated in the groupedstats
+uint64 errors = 8; // count of error spans aggregated in the groupedstats
+uint64 duration = 9; // total duration in nanoseconds of spans aggregated in the bucket
+bytes okSummary = 10; // ddsketch summary of ok spans latencies encoded in protobuf
+bytes errorSummary = 11; // ddsketch summary of error spans latencies encoded in protobuf
+bool synthetics = 12; // set to true on spans generated by synthetics traffic
+uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
+```
+
+So it is basically a group of various metrics that can be represented in Vector (Sketches are support since this
+[PR][sketch-pr]). In the proto definition sketches are stored as unstructured bytes slices, but those
+[fields][call-to-toproto] are filled with a [protobuf encoded ddsketch][ddsktech-proto]. Given that sketches in Vector
+are also [heavily based on ddsketch][vector-sketch], APM stats sketches can be converted to/from the Vector internal
+representation without incuring too much accuracy losses, but this would require significant work to implement those
+conversion.
```

Comment:
```suggestion
As you can see, it is a group of various metrics that can be represented in Vector as such (Sketches are supported since this
[PR][sketch-pr]). In the proto definition sketches are stored as unstructured bytes slices, but those
[fields][call-to-toproto] are filled with a [protobuf encoded ddsketch][ddsktech-proto]. Given that sketches in Vector
are also [heavily based on ddsketch][vector-sketch], APM stats sketches can be converted to/from the Vector internal
representation without incurring too much accuracy loss, but this would require significant work to implement that
conversion.
```